hail/python/hail/ir/matrix_reader.py

```diff
@@ -20,18 +22,46 @@ def __eq__(self, other):
 
 
 class MatrixNativeReader(MatrixReader):
-    @typecheck_method(path=str)
-    def __init__(self, path):
+    @typecheck_method(path=str,
+                      intervals=nullable(sequenceof(anytype)),
+                      filter_intervals=bool)
+    def __init__(self, path, intervals, filter_intervals):
+        if intervals is not None:
+            t = hl.expr.impute_type(intervals)
+            if not isinstance(t, hl.tarray) and not isinstance(t.element_type, hl.tinterval):
+                raise TypeError("'intervals' must be an array of tintervals")
+            pt = t.element_type.point_type
+            if isinstance(pt, hl.tstruct):
+                self._interval_type = t
+            else:
+                self._interval_type = hl.tarray(hl.tinterval(hl.tstruct(__point=pt)))
+
         self.path = path
+        self.intervals = hl.eval(intervals) if intervals is not None else None
```

Comment:
yeah, we definitely can't call `hl.eval()` in IR initialization. If we want to recons the IR, suddenly that's a super slow operation.